server/v2/commands.go

```diff
@@ -11,27 +11,46 @@ import (
 
 	"github.com/spf13/cobra"
 
+	"cosmossdk.io/core/transaction"
 	"cosmossdk.io/log"
+	"github.com/spf13/viper"
 )
 
-func Commands(logger log.Logger, homePath string, modules ...ServerModule) (CLIConfig, error) {
-	if len(modules) == 0 {
-		// TODO figure if we should define default modules
+const flagHome = "home"
+
+type App[T transaction.Tx] struct {
+	Application Application[T]
+	Store       any
+}
+
+type AppCreator[T transaction.Tx] func(*viper.Viper, log.Logger) App[T]
+
+func Commands(rootCmd *cobra.Command, newApp AppCreator[transaction.Tx], logger log.Logger, components ...ServerComponent[transaction.Tx]) (CLIConfig, error) {
+	if len(components) == 0 {
+		// TODO figure if we should define default components
 		// and if so it should be done here to avoid uncessary dependencies
-		return CLIConfig{}, errors.New("no modules provided")
+		return CLIConfig{}, errors.New("no components provided")
 	}
 
-	v, err := ReadConfig(filepath.Join(homePath, "config"))
-	if err != nil {
-		return CLIConfig{}, fmt.Errorf("failed to read config: %w", err)
-	}
+	server := NewServer(logger, components...)
+	// Write default config for each server module
+	flags := server.StartFlags()
 
-	server := NewServer(logger, modules...)
 	startCmd := &cobra.Command{
 		Use:   "start",
 		Short: "Run the application",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			if err := v.BindPFlags(cmd.Flags()); err != nil { // the server modules are already instantiated here, so binding the flags is useless.
+			v := GetViperFromCmd(cmd)
+			l := GetLoggerFromCmd(cmd)
+
+			app := newApp(v, l)
```

Comment:
the init and the newApp should be moved last, after we binded all the startflags